core/corehttp/hostname.go

```diff
@@ -0,0 +1,301 @@
+			// 3. does DNSLink record exist in DNS?
+			if !cfg.Gateway.NoDNSLink && isDNSLinkRequest(r, n) {
+				// rewrite path and handle as DNSLink
+				r.URL.Path = "/ipns/" + stripPort(r.Host) + r.URL.Path
+				childMux.ServeHTTP(w, r)
+				return
+			}
+
+			// else, treat it as an old school gateway, I guess.
+			childMux.ServeHTTP(w, r)
+		})
+		return childMux, nil
+	}
+}
+
+// isDNSLinkRequest returns bool that indicates if request
+// should return data from content path listed in DNSLink record (if exists)
+func isDNSLinkRequest(r *http.Request, n *core.IpfsNode) bool {
+	ctx, cancel := context.WithCancel(n.Context())
+	defer cancel()
+
+	fqdn := stripPort(r.Host)
+	if len(fqdn) > 0 && isd.IsDomain(fqdn) {
+		// Confirm DNSLink was not disabled for the fqdn or globally
+		name := "/ipns/" + fqdn
+		// check if DNSLink exists
+		_, err := n.Namesys.Resolve(ctx, name, nsopts.Depth(1))
+		if err == nil || err == namesys.ErrResolveRecursion {
+			return true
+		}
+	}
+	return false
+}
+
+func isSubdomainNamespace(ns string) bool {
+	switch ns {
+	case "ipfs", "ipns", "p2p", "ipld", "api":
+		return true
+	default:
+		return false
+	}
+}
+
+// Parses Host header of a subdomain-based URL and returns it's components
+// Note: hostname is host + optional port
+func parseSubdomains(hostHeader string) (hostname, ns, rootID string, ok bool) {
+	parts := subdomainGatewayRegex.FindStringSubmatch(hostHeader)
+	if len(parts) < 4 || !isSubdomainNamespace(parts[2]) {
+		return "", "", "", false
+	}
+	return parts[3], parts[2], parts[1], true
+}
+
+// Converts a hostname/path to a subdomain-based URL, if applicable.
+func toSubdomainURL(hostname, path string, query string) (url string, ok bool) {
+	parts := strings.SplitN(path, "/", 4)
+
+	var ns, rootID, rest string
+	switch len(parts) {
+	case 4:
+		rest = parts[3]
+		fallthrough
+	case 3:
+		ns = parts[1]
+		rootID = parts[2]
+	default:
+		return "", false
+	}
+
+	if !isSubdomainNamespace(ns) {
+		return "", false
+	}
+
+	// add prefix if query is present
+	if query != "" {
+		query = "?" + query
+	}
+
+	if ns == "api" || ns == "p2p" {
+		// API and P2P proxy use the same paths on subdomains:
+		// api.hostname/api/.. and p2p.hostname/p2p/..
+		return fmt.Sprintf(
+			"http://%s.%s/%s/%s/%s%s",
+			ns,
+			hostname,
+			ns,
+			rootID,
+			rest,
+			query,
+		), true
+	}
+
+	if rootCid, err := cid.Decode(rootID); err == nil {
+		multicodec := rootCid.Type()
+
+		// CIDs in IPNS are expected to have libp2p-key multicodec.
+		// We ease the transition by fixing multicodec on the fly:
+		// https://github.com/ipfs/go-ipfs/issues/5287#issuecomment-492163929
+		if ns == "ipns" && multicodec != cid.Libp2pKey {
+			multicodec = cid.Libp2pKey
+		}
+
+		// if object turns out to be a valid CID,
+		// ensure text representation used in subdomain is CIDv1 in Base32
+		// https://github.com/ipfs/in-web-browsers/issues/89
+		rootID = cid.NewCidV1(multicodec, rootCid.Hash()).String()
```

Comment:
We should explicitly pick the pace here instead of assuming the default is base32.